### PR TITLE
QgsProject is no longer a singleton

### PR DESCRIPTION
**Pull request by m-kuhn**
use QgsApplication::activeProject instead

## Description

QgsProject being a singleton has over time led to many workarounds that make the codebase hard to maintain in a sustainable way.

This pull request proposes to get rid of the problem by delegating the project management to a new method on application level.
`QgsApplication::activeProject()` right now pretty much does the same thing as `QgsProject::instance()` did before. There is one project created on startup and deleted on shutdown of the application.
By itself this offers no risks. But, once someone starts using `setActiveProject` from plugins or other code, there's a certain risk involved, that existing code is not prepared for projects "just disappearing". This scenario will need to be carefully tested before we introduce calls to `setActiveProject` somewhere in code we ship.

Other open questions (which can be addressed in the future as well):

 * Should we have a list of open projects, owned by the application, of which one is the active one? I.e. have a project registry.
 * Do we want to be able to run without an activeProject()

### CODE REVIEW

**Comment by luipir** [comment]
is it an api break?

**Comment by m-kuhn** [comment]
doesn't look like one to me, but you may convince me it is one :)

File: src/core/qgsapplication.cpp [hunk 1]
```diff
@@ -289,6 +292,9 @@ void QgsApplication::init( QString profileFolder )
 
 QgsApplication::~QgsApplication()
 {
+  if ( mApplicationMembers->mActiveProject && mApplicationMembers->mActiveProject->parent() == this )
+    delete mApplicationMembers->mActiveProject;
```
Comment (m-kuhn) [inline src/core/qgsapplication.cpp]: Note to self, that's quite redundant

**Comment by luipir** [comment]
I'm not agaist the PR, I'm just saying we are asking to adapt future code to use this new way to access the project. How much time we should maintain the instance() fake method?

**Comment by m-kuhn** [comment]
Until QGIS 4.0

File: src/core/qgsapplication.cpp (hunk 1 above)
Comment (elpaso) [inline src/core/qgsapplication.cpp]: Just asking: is there any chance to have automatic memory management such as std::unique_ptr for `mActiveProject` ?

Comment (m-kuhn) [reply src/core/qgsapplication.cpp]: No, mixing QObject parents and unique_ptrs is a bad idea

**Comment by nyalldawson** [comment]
Just wondering... Is this the ideal approach? I'd have thought we want to move this to iface and have it as an app only thing (in the ideal world).

**Comment by m-kuhn** [comment]
The advantage would be, the whole codebase would be forced to pass the project always through the whole call stack. The disadvantage would be, the whole codebase would be forced to pass the project always through the whole call stack.

**Comment by m-kuhn** [comment]
Working on the nasty details here. QgsTaskManager has a dependency on the QgsProject to track removal of layers. I don't think we can send that through the main app. So at least a list of projects needs to be present inside QgsApplication.
On the other hand, this does not per-se mean that the activeProject needs to live in the same place.

Looking at the fact, that this is triggered by https://github.com/qgis/QGIS/pull/7244, where it's about resolving value relations, that should not be done using a QgsApplication based activeProject... So I withdraw this pull request for an iface based implementation.

**Pull request closed without merging**